### PR TITLE
Implemented Lanczos Algorithm

### PR DESCRIPTION
**Pull request by felix642**
This PR contains : 
- Implementation of the Lanczos Algorithm
- Associated documentation
- Associated test to validate the algorithm

The Gram-Schmidt algorithm is used to reduce the numerical instability of Lanczos.

closes felix642/1chipML#19

### CODE REVIEW

**Review by bisson2000 — APPROVED** [review]
lgtm

File: src/lanczos.c [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    vectorScale(tmpVector, vectorLength, dotProduct);
+    vectorSubstract(vectorToChange, tmpVector, vectorLength);
+  }
+  lanczos_real norm = computeNorm(vectorToChange, vectorLength);
+  vectorScale(vectorToChange, vectorLength, 1.0 / norm);
+}
+
+static void getRandomUnitVector(lanczos_real* vector, uint_least8_t dim) {
+  for (uint_least8_t i = 0; i < dim; ++i) {
+    vector[i] = rand() % 10;
```
Comment (GM07) [inline src/lanczos.c:95]: Please use rand from congruential random generator.

File: src/lanczos.c [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    // Store q0 now becomes q1 and q1 is computed using the equation
+    // q1 = v / beta
+    //
+    // Special case where beta is 0
+    // In that case q1 should be a new randomly generated vector
+    // orthogonal to all previous vectors in vTranspose
+    memcpy(q0, q1, dim * sizeof(lanczos_real));
+    memcpy(q1, v, dim * sizeof(lanczos_real));
+    if (beta >= -0.00001 && beta <= 0.00001) {
```
Comment (GM07) [inline src/lanczos.c:156]: fabs and epsilon and we need to check memcpy

**Review by nicholaslegrand102 — APPROVED** [review]

**Review by jmsellier — APPROVED** [review]
GFM!

**Pull request merged**